examples/seismic/viscoelastic/operators.py

```diff
@@ -0,0 +1,309 @@
+import numpy as np
+import sympy as sp
+
+from devito import Eq, Operator, TimeFunction, NODE
+from examples.seismic import PointSource, Receiver
+
+
+def stress_fields(model, save, space_order):
+    """
+    Create the TimeFunction objects for the stress fields in the viscoelastic formulation.
+    """
+    if model.grid.dim == 2:
+        x, z = model.space_dimensions
+        stagg_xx = stagg_zz = NODE
+        stagg_xz = (x, z)
+        # Create symbols for forward wavefield, source and receivers
+        txx = TimeFunction(name='txx', grid=model.grid, staggered=stagg_xx, save=save,
+                           time_order=1, space_order=space_order)
+        tzz = TimeFunction(name='tzz', grid=model.grid, staggered=stagg_zz, save=save,
+                           time_order=1, space_order=space_order)
+        txz = TimeFunction(name='txz', grid=model.grid, staggered=stagg_xz, save=save,
+                           time_order=1, space_order=space_order)
+        tyy = txy = tyz = None
+    elif model.grid.dim == 3:
+        x, y, z = model.space_dimensions
+        stagg_xx = stagg_yy = stagg_zz = NODE
+        stagg_xz = (x, z)
+        stagg_yz = (y, z)
+        stagg_xy = (x, y)
+        # Create symbols for forward wavefield, source and receivers
+        txx = TimeFunction(name='txx', grid=model.grid, staggered=stagg_xx, save=save,
+                           time_order=1, space_order=space_order)
+        tzz = TimeFunction(name='tzz', grid=model.grid, staggered=stagg_zz, save=save,
+                           time_order=1, space_order=space_order)
+        tyy = TimeFunction(name='tyy', grid=model.grid, staggered=stagg_yy, save=save,
+                           time_order=1, space_order=space_order)
+        txz = TimeFunction(name='txz', grid=model.grid, staggered=stagg_xz, save=save,
+                           time_order=1, space_order=space_order)
+        txy = TimeFunction(name='txy', grid=model.grid, staggered=stagg_xy, save=save,
+                           time_order=1, space_order=space_order)
+        tyz = TimeFunction(name='tyz', grid=model.grid, staggered=stagg_yz, save=save,
+                           time_order=1, space_order=space_order)
+
+    return txx, tyy, tzz, txy, txz, tyz
+
+
+def memory_variables(model, save, space_order):
```

Comment:
auxiliary/attenuation variables, memory is a weird name for it.

Reply:
Some people call them memory variables i.e. Bohlen (2002) among others. Anything, within reason, is fine with me though.